src/components/pages/data-export-page/data-export-page.tsx

```diff
@@ -0,0 +1,190 @@
+  Paper,
+} from '@material-ui/core';
+import {
+  DatePicker,
+  MuiPickersUtilsProvider,
+} from '@material-ui/pickers';
+import { Moment } from 'moment';
+import React, {
+  useState,
+  useCallback,
+} from 'react';
+import { useSelector } from 'react-redux';
+import { ApiError } from '../../../models/api-response';
+import { UserSelector } from '../../../selectors/user.selector';
+import { downloadFile } from '../../../utility/download';
+import { getLineCount } from '../../../utility/string-utils';
+import {
+  AlertDialog,
+  AlertDialogProps,
+} from '../../alert-dialog/alert-dialog';
+import { ButtonWithSpinner } from '../../buttons/button-with-spinner';
+import { LinearProgressWithPercent } from '../../linear-progress-with-label/linear-progress-with-label';
+import PageHeader from '../../page-header/page-header';
+import useStyles from './data-export-page.styles';
+
+export const DataExportPage = () => {
+  const classes = useStyles();
+
+  const [startDate, setStartDate] = useState<Moment | null>(null);
+  const [endDate, setEndDate] = useState<Moment | null>(null);
+  const [isExportLoading, setIsExportLoading] = useState(false);
+  const [exportProgress, setExportProgress] = useState(0);
+  const [alertDialogProps, setAlertDialogProps] = useState<AlertDialogProps>({ open: false });
+
+  const orgId = useSelector(UserSelector.orgId);
+
+  const runExport = useCallback(async () => {
+    setExportProgress(0);
+
+    // Use fetch() here since axios doesn't support streaming.
+    const queryParams = [];
+    if (startDate) {
+      queryParams.push(`startDate=${startDate.toISOString()}`);
+    }
+    if (endDate) {
+      queryParams.push(`endDate=${endDate.toISOString()}`);
+    }
+    const response = await fetch(new Request(`api/org/${orgId}/export?${queryParams.join('&')}`, {
+      headers: {
+        'Content-Type': 'application/octet-stream',
+      },
+    }));
+
+    if (response.status !== 200) {
+      let message: string | undefined;
+      try {
+        const error = await response.json() as ApiError;
+        message = error.errors[0].message;
+      } finally {
+        if (!message) {
+          message = response.statusText;
+        }
+      }
+
+      throw new Error(`Failed to export data: ${message}`);
+    }
+
+    const totalHits = parseInt(response.headers.get('total-hits')!);
+
+    const reader = response.body!.getReader();
+    const textDecoder = new TextDecoder();
+    let dataStr = '';
+    let hitsRead = -1; // Offset for csv header line.
+    let chunk = await reader.read();
+    while (!chunk.done) {
+      const chunkStr = textDecoder.decode(chunk.value);
+      dataStr += chunkStr;
+
+      hitsRead += getLineCount(chunkStr);
+      setExportProgress((hitsRead / totalHits) * 100);
+
+      chunk = await reader.read();
+    }
+
+    downloadFile(dataStr, 'export', 'csv');
+  }, [orgId, startDate, endDate]);
+
+  const handleExportButtonClick = () => {
+    setIsExportLoading(true);
+
+    runExport()
```

Comment:
can we change this to use `await` in a `try/catch/finally` block to be more consistent with the rest of the codebase?

Reply:
Whoops, for some reason I was thinking that function couldn't use async.